### DISCUSSION TITLE
Number of remaining API calls?

### DISCUSSION THREAD

**Discussion by andreafailla** [MarshalX/atproto#286] [Questions] [answered]
Hello, I'm building something that requires to download posts from an actor's feed. Is there any way to see the number of API calls I can still make before a TooManyRequests is thrown?
Thanks!

**Comment by MarshalX** [comment]
Hello, I think it duplicates https://github.com/MarshalX/atproto/discussions/167. I wanna provide such API but I'm not sure of the best implementation. Saving the last response headers to the client could be problematic in async version. Maybe adding some kind of callback which will be called on plain response with headers is good enough. Wdyt? Do you have examples of any other Python libs that provide such functionality?

**Reply by andreafailla** [reply]
Hello; yep this is probably a duplicate – my bad, sorry! 
So, the [Mastodon API Python wrapper](https://github.com/halcy/Mastodon.py) ([docs](https://mastodonpy.readthedocs.io/en/stable/01_general.html#rate-limiting)) can tell how many calls you have left. Not sure of the details in async.

However, I guess that most of the usecases of needing rate limit info is to avoid hitting such limit. One way to circumvent the issue would be to just pause/queue all next requests until 'RateLimit-Reset' when 'RateLimit-Remaining' is 1.
Again, unfortunately i'm not familiar with async so mine are just wild guesses.

**Comment by MarshalX** [comment, marked as answer]
upd from 2026:

429 now raises a dedicated exception

```python
except RateLimitExceededError as e:
    e.limit, e.remaining, e.reset_at, e.policy  # the ratelimit-* headers
    e.retry_after. # retry-after, in seconds
```

---

Hello. I prepared simple example of override method that gives you access to response headers (which contains rate limits).

```python
class RateLimitedClient(Client):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

        self._limit = self._remaining = self._reset = None

    def get_rate_limit(self):
        return self._limit, self._remaining, self._reset

    def _invoke(self, *args, **kwargs):
        response = super()._invoke(*args, **kwargs)

        self._limit = response.headers.get('RateLimit-Limit')
        self._remaining = response.headers.get('RateLimit-Remaining')
        self._reset = response.headers.get('RateLimit-Reset')

        return response
```

Usage:
```python
client = RateLimitedClient()
client.login(os.environ['USERNAME'], os.environ['PASSWORD'])

post = client.get_timeline()
print(client.get_rate_limit())
```
